### PR TITLE
TEL - Right Hand of the Emperor

### LINKED ISSUE #3528: TEL - Right Hand of the Emperor
image

### PR DESCRIPTION
**Pull request by marcsoiferman-cm**
Implementation and tests for Right Hand of the Emperor (closes #3528)

Card operates based on the following rule interpretations:
1) You can only choose bowed bushi you control (the resolution has to be able to effect the characters you target)
2) You can choose to target 0 characters, which causes the card to be placed in your deck and ready no one.
3) You can play the card if you have no eligible targets, which feeds into 2.

### CODE REVIEW

File: server/game/cards/10-TEL/RightHandOfTheEmperor.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            target: {
+                mode: TargetModes.MaxStat,
+                activePromptTitle: 'Choose characters',
+                cardStat: card => card.getCost(),
+                maxStat: () => 6,
+                numCards: 0,
+                cardType: CardTypes.Character,
+                controller: Players.Self,
+                cardCondition: card => card.hasTrait('bushi') && card.bowed,
+                gameAction: AbilityDsl.actions.sequential([AbilityDsl.actions.ready(),
```
Comment (jeremylarner) [inline server/game/cards/10-TEL/RightHandOfTheEmperor.js]: this should be `multiple` - also can you format in line with other cards? :)

Comment (marcsoiferman-cm) [reply server/game/cards/10-TEL/RightHandOfTheEmperor.js]: Switched to multiple.  I think the formatting is better now, looked at a few examples of other cards.

File: server/game/cards/10-TEL/RightHandOfTheEmperor.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            title: 'Ready characters',
+            target: {
+                mode: TargetModes.MaxStat,
+                activePromptTitle: 'Choose characters',
+                cardStat: card => card.getCost(),
+                maxStat: () => 6,
+                numCards: 0,
+                cardType: CardTypes.Character,
+                controller: Players.Self,
+                cardCondition: card => card.hasTrait('bushi') && card.bowed,
```
[the remark this answers is not in the document]
Comment (marcsoiferman-cm) [reply server/game/cards/10-TEL/RightHandOfTheEmperor.js]: Strangely enough the engine was letting me target readied characters, which is why I put this condition in.

**Comment by marcsoiferman-cm** [comment]
I don't like 
`numCards: 99`

But numCards: 0 doesn't seem to work with optional cost and TargetModes.MaxStat.  It wouldn't let me target anyone.

**Comment by jeremylarner** [comment]
There may well be a bug with the card selectors, they don't get tested much :P I'm happy to take a look at it if you're still having problems after restructuring your game actions

File: server/game/cards/10-TEL/RightHandOfTheEmperor.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremylarner) [reply server/game/cards/10-TEL/RightHandOfTheEmperor.js]: See my comment below

File: server/game/cards/10-TEL/RightHandOfTheEmperor.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                mode: TargetModes.MaxStat,
+                activePromptTitle: 'Choose characters',
+                cardStat: card => card.getCost(),
+                maxStat: () => 6,
+                numCards: 99, //numCards: 0 doesn't seem to work with optional cost and TargetModes.MaxStat
+                optional: true,
+                cardType: CardTypes.Character,
+                controller: Players.Self,
+                cardCondition: card => card.hasTrait('bushi') && card.bowed,
+                gameAction: AbilityDsl.actions.multiple([
```
[the remark this answers is not in the document]
Comment (marcsoiferman-cm) [reply server/game/cards/10-TEL/RightHandOfTheEmperor.js]: Updated the gameAction to be split out of the target block.  numCards: 0 still doesn't work though.  Going to continue using 99 as a workaround.

**Comment by jeremylarner** [comment]
I realise you have a million tests on this card already, but can I ask you to add one more which checks that playing it from discard doesn't increase its costs when Utaku Tetsuko is attacking (or that Akodo Toturi doesn't stop it from being played from discard, I guess)?

I suspect this test will probably fail, at which point I will try and debug the issues you're having for both this and the card selectors

**Comment by jeremylarner** [comment]
Sorry, I have another test request - can you test to see whether GoH or MoGT stop it from being played from discard?

**Comment by jeremylarner** [comment]
Needs #3602

**Comment by marcsoiferman-cm** [comment]
I added the condition that the target must be bowed due to the following rules block:

> A card is not an eligible target for an ability if the resolution of that ability's effect could not affect the target at all, unless it is a referential target. (For example, a bowed character cannot be chosen as the target for an ability that reads "Action: Choose a character – bow that character.")

Does this block not apply here?

**Comment by marcsoiferman-cm** [comment]
Needs #3626

**Comment by jeremylarner** [comment]
The engine knows about targeting requirements in the rules, so you shouldn't need to put anything in the card implementation beyond what it says on the card.  So unless the card says 'Choose a bowed character', you shouldn't need to explicitly check that the character is bowed.

This PR seems to have become massive - is it actually just the card implementation and the test?  If so, would you mind closing this and creating a new PR?  If there are other changes, I think we can manage - it's just difficult to remember what you've actually changed...

**Comment by jeremylarner** [comment]
Resubmitted in #3639

**Pull request closed without merging**